sdk/samples/iot/hub/src/paho_iot_hub_pnp_example.c

```diff
@@ -0,0 +1,1020 @@
+                &json_builder, desired_temp_property_name, temp_value, 200, version, AZ_SPAN_FROM_STR("success"))))
+    {
+      return rc;
+    }
+  }
+  az_span json_payload = az_json_builder_get_json(&json_builder);
+
+  printf("Payload: %.*s\n", az_span_size(json_payload), (char*)az_span_ptr(json_payload));
+
+  // Publish the reported property payload to IoT Hub
+  rc = mqtt_publish_message(reported_property_topic, json_payload, 0);
+
+  return rc;
+}
+
+// Parse the desired temperature property from the incoming JSON
+static az_result parse_twin_desired_temperature_property(
+    az_span twin_payload_span,
+    bool is_twin_get,
+    uint32_t* parsed_value,
+    int32_t* version_number)
+{
+  az_json_parser jp;
+  bool desired_found = false;
+  AZ_RETURN_IF_FAILED(az_json_parser_init(&jp, twin_payload_span, NULL));
+  AZ_RETURN_IF_FAILED(az_json_parser_next_token(&jp));
+  if (jp.token.kind != AZ_JSON_TOKEN_BEGIN_OBJECT)
+  {
+    return AZ_ERROR_PARSER_UNEXPECTED_CHAR;
+  }
+
+  if (is_twin_get)
+  {
+    // If is twin get payload, we have to parse one level deeper for "desired" wrapper
+    AZ_RETURN_IF_FAILED(az_json_parser_next_token(&jp));
+    while (jp.token.kind != AZ_JSON_TOKEN_END_OBJECT)
+    {
+      if (az_json_token_is_text_equal(&jp.token, desired_property_name))
+      {
+        desired_found = true;
+        AZ_RETURN_IF_FAILED(az_json_parser_next_token(&jp));
+        break;
+      }
+      else
+      {
+        // else ignore token.
+        AZ_RETURN_IF_FAILED(az_json_parser_skip_children(&jp));
+      }
+
+      AZ_RETURN_IF_FAILED(az_json_parser_next_token(&jp));
+    }
+  }
+  else
+  {
+    desired_found = true;
+  }
+
+  if (!desired_found)
+  {
+    printf("Desired property object not found in twin");
+    return AZ_ERROR_ITEM_NOT_FOUND;
+  }
+
+  if (jp.token.kind != AZ_JSON_TOKEN_BEGIN_OBJECT)
+  {
+    return AZ_ERROR_PARSER_UNEXPECTED_CHAR;
+  }
+  AZ_RETURN_IF_FAILED(az_json_parser_next_token(&jp));
+
+  bool temp_found = false;
+  bool version_found = false;
+  while (!(temp_found && version_found) || (jp.token.kind != AZ_JSON_TOKEN_END_OBJECT))
+  {
+    if (az_json_token_is_text_equal(&jp.token, desired_temp_property_name))
+    {
+      AZ_RETURN_IF_FAILED(az_json_parser_next_token(&jp));
+      AZ_RETURN_IF_FAILED(az_json_token_get_uint32(&jp.token, parsed_value));
+      temp_found = true;
+    }
+    else if (az_json_token_is_text_equal(&jp.token, AZ_SPAN_FROM_STR("$version")))
+    {
+      AZ_RETURN_IF_FAILED(az_json_parser_next_token(&jp));
+
+      AZ_RETURN_IF_FAILED(az_json_token_get_uint32(&jp.token, (uint32_t*)version_number));
+      version_found = true;
+    }
+    else
+    {
+      // else ignore token.
+      AZ_RETURN_IF_FAILED(az_json_parser_skip_children(&jp));
+    }
+    AZ_RETURN_IF_FAILED(az_json_parser_next_token(&jp));
+  }
+  
+  if(temp_found && version_found)
+  {
+    return AZ_OK;
+  }
+
+  return AZ_ERROR_ITEM_NOT_FOUND;
+}
+
+static bool update_device_temp(int32_t temp)
```

Comment:
>update_device_temp [](start = 12, length = 18)

Having this function return true if max temp updated and false otherwise seems to violate principle of least astonishment.  I'd expect an update_device_temp() failing to mean that we couldn't take the value (e.g. set to 50000 degrees desired).  Recommend having the max_temp changed be returned in a parameter so intent is clearer.

Reply:
good point. will change